subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultMavenRepositoryContentDescriptor.java

```diff
@@ -65,4 +66,25 @@ public void snapshotsOnly() {
         }
         return filter;
     }
+
+    @Override
+    public RepositoryContentDescriptorInternal asMutableCopy() {
+        DefaultMavenRepositoryContentDescriptor copy = new DefaultMavenRepositoryContentDescriptor(getRepositoryNameSupplier());
+        if (getIncludedConfigurations() != null) {
+            copy.setIncludedConfigurations(Sets.newHashSet(getIncludedConfigurations()));
+        }
+        if (getExcludedConfigurations() != null) {
+            copy.setExcludedConfigurations(Sets.newHashSet(getExcludedConfigurations()));
+        }
+        if (getIncludeSpecs() != null) {
+            copy.setIncludeSpecs(Sets.newHashSet(getIncludeSpecs()));
+        }
+        if (getExcludeSpecs() != null) {
+            copy.setExcludeSpecs(Sets.newHashSet(getExcludeSpecs()));
+        }
+        if (getRequiredAttributes() != null) {
+            copy.setRequiredAttributes(Maps.newHashMap(getRequiredAttributes()));
```

Comment:
I think a deeper copy needs to occur here.

1. Is `Attribute` mutable?
2. The value in this map is itself a `Set<Object>`. Doesn't the value for every key need to be copied to a new `Set`?

Reply:
1. No, it is not.
2. In theory, but in practice once that is realised, it is effectively immutable. If it was mutable, subsequent invocation of the content filter for a single repo could lead to inconsistent filtering. So by design it has to be effectively immutable.